tests/factories.py

```diff
@@ -0,0 +1,121 @@
+import base64
+import secrets
+
+from jobrunner import job_executor
+from jobrunner.models import Job, JobRequest, SavedJobRequest, State
+from jobrunner.lib.database import insert
+from jobrunner.manage_jobs import JobError
+
+
+JOB_REQUEST_DEFAULTS = {
+    "repo_url": "repo",
+    "commit": "commit",
+    "requested_actions": [],
+    "cancelled_actions": [],
+    "workspace": "workspace",
+    "database_name": "full",
+    "original": {
+        "created_by": "testuser",
+    },
+}
+
+
+JOB_DEFAULTS = {
+    "action": "action_name",
+    "repo_url": "opensafely/study",
+    "workspace": "workspace",
+    "requires_outputs_from": [],
+    "run_command": "python myscript.py",
+    "output_spec": {},
+}
+
+
+def job_request_factory(**kwargs):
+    if "id" not in kwargs:
+        kwargs["id"] = base64.b32encode(secrets.token_bytes(10)).decode("ascii").lower()
+
+    values = JOB_REQUEST_DEFAULTS.copy()
```

Comment:
Don't you need a `deepcopy()` here (and with `JOB_DEFAULTS` below) because you've got mutable values in the dicts (`requested_actions`, `output_spec` etc)?

Reply:
Hmm, I do, yes, but I don't actually modify them

Reply:
fixed